File: ts/project/.sdk/tm/clojure/test/sdk/test/omni.clj
;; ProjectName SDK corpus test runner: the vendored @voxgig/omni engine driven
;; through its NATIVE API (`voxgig.omni.runner/make-runner`), presented to the
;; corpus suites in the runner shape they already use (`:spec`, `:runset`,
;; `:runsetflags`, `:client`). No compatibility shim is vendored: the adapter
;; below IS the whole bridge, per language, per the vendor-tag rollout
;; (docs/design/vendor-tag-rollout.md, Decision 4). It is the clojure peer of
;; tm/lua/test/omni.lua, tm/rb/test/omni.rb and tm/java/test/OmniResolver.java.
;;
;; Four local decisions, all required:
;;
;; 1. TWO VALUE MODELS, converted at the subject boundary. omni's clojure port
;;    speaks IMMUTABLE clojure data - `ismap` is `map?`, `islist` is `vector?`,
;;    absence is the `::u/absent` keyword. The SDK and the vendored struct
;;    speak MUTABLE JVM collections (java.util.LinkedHashMap / ArrayList),
;;    because struct's contract is that a utility MUTATES the node it is
;;    given. Neither recognises the other, so `wrap-subject` converts every
;;    argument on the way in (`->struct`) and the result on the way out
;;    (`->omni`).
;;
;; 2. NO-VALUE ARITY (the zero-argument correction the dynamic ports carry).
;;    An entry with no `in` must reach the subject as NO value, not as one
;;    null: the corpus has both `{in: null, out: <T_null>}` and
;;    `{out: <T_noval>}`, and a port that collapses them silently passes one
;;    of the two. omni models the absence as its ABSENT sentinel, and here it
;;    becomes a genuine ZERO-ARITY call whenever the subject HAS a zero-arity
;;    arm - `s/typify` is handed to the corpus bare, so `(typify)` answers
;;    T_noval where `(typify nil)` answers T_null.
;;
;;    A subject with no zero-arity arm gets `nil` instead. That fallback is
;;    not a convenience: struct's NOARG is TYPIFY's marker, not a general
;;    no-value, so handing it to a one-arity subject is simply a wrong
;;    argument - `(isempty NOARG)` answers false where the corpus wants the
;;    true that `(isempty nil)` gives, and `(clone NOARG)` answers a sentinel
;;    where it wants null. The corpus caught both. It is also the LIVE path
;;    for `minor.stringify` and `minor.pathify`: their corpus subjects are
;;    one-arity wrappers, so they take the nil and make struct's own no-arg
;;    call themselves.
;;
;;    Which arm a subject has is READ OFF THE FUNCTION (`accepts-no-args?`),
;;    never discovered by calling it and catching ArityException - see the
;;    note on that function.
;;
;; 2a. WHAT RAN IS COUNTED, not what the corpus declared. A section's pass
;;    count comes from `subject-calls`, incremented inside the subject
;;    wrapper - the one place only the engine can reach - and `runsetflags`
;;    returns the delta across the set. A count taken from the section's own
;;    `set` prints the same number whether the engine ran or was never
;;    called, which makes it evidence of nothing.
;;
;; 3. THE CTX CHANNEL (`live-map`). omni snapshots `entry.ctx` BEFORE calling
;;    the subject, then matches against that snapshot afterwards. In every
;;    mutable-language port the snapshot and the argument are the same live
;;    object, so a utility that writes `ctx.spec.headers.authorization` is
;;    visible to `match: {ctx: ...}` (9 primary entries turn on exactly that).
;;    Clojure's data is immutable, so the snapshot would freeze the
;;    pre-call state and every such assertion would read null. `contextify`
;;    therefore hands omni a LIVE VIEW: a type that satisfies `map?` (so
;;    `u/ismap`, `u/getpath`, `u/deepequal` and `u/stringify` all accept it)
;;    but reads through to the mutable map the subject actually mutates. The
;;    port's own `:runsetflags-args` channel carries the ARGUMENT writeback;
;;    this carries the ctx one, which `runsetflags-args` cannot reach.
;;
;; 4. LIVE OBJECTS CROSS BY REFERENCE. A value that belongs to neither model -
;;    the provider, an SDK client, a Context atom - passes through the
;;    converters as a leaf, so subjects keep real object identity (and so a
;;    cyclic SDK object is never walked).

(ns sdk.test.omni
  (:require [voxgig.omni.runner :as runner]
            [voxgig.omni.util :as u]
            [voxgig.struct :as vs]
            [sdk.core :as core]
            [sdk.client :as client]
            [clojure.string :as str])
  (:import [java.util LinkedHashMap ArrayList List Map IdentityHashMap]))


;; omni's sentinels and error predicate, re-exported so the corpus suites and
;; the smoke test never reach into the vendored namespaces themselves.
(def NULLMARK u/NULLMARK)
(def UNDEFMARK u/UNDEFMARK)
(def EXISTSMARK u/EXISTSMARK)
(def ABSENT u/ABSENT)

(defn omni-error? [err] (runner/omni-error? err))
(defn deep-equal? [a b] (u/deepequal a b))
(defn stringify [val] (u/stringify val))


;; --------------------------------------------------------------------------
;; Decision 4: the live-object marker.
;; --------------------------------------------------------------------------

(def ^:private LIVEMETA ::live)

(defn- mark-live [val] (vary-meta val assoc LIVEMETA true))

(defn- live? [val]
  (boolean (and (instance? clojure.lang.IObj val) (get (meta val) LIVEMETA))))


;; --------------------------------------------------------------------------
;; Decision 3: a live-reading omni map over a mutable struct map.
;; --------------------------------------------------------------------------

(defprotocol ILiveMap
  (live-backing [this] "The mutable struct map this view reads through to."))

(declare ->omni ->struct)

;; A `map?` that is not a map: every read re-converts the mutable backing map,
;; so omni sees the state as it is AT MATCH TIME rather than at call time.
;; `extra` carries keys omni assoc'd on afterwards (it attaches the active
;; client under "client"), which must not be written into the SDK's own node.
(defn- live-map [^Map backing extra]
  (let [snap (fn [] (merge (->omni backing) extra))]
    (reify
      ILiveMap
      (live-backing [_] backing)

      clojure.lang.ILookup
      (valAt [_ k] (get (snap) k))
      (valAt [_ k nf] (get (snap) k nf))

      clojure.lang.Associative
      (containsKey [_ k] (contains? (snap) k))
      (entryAt [_ k] (find (snap) k))
      (assoc [_ k v] (live-map backing (assoc extra k v)))

      clojure.lang.IPersistentMap
      (assocEx [_ k v] (live-map backing (assoc extra k v)))
      (without [_ k] (live-map backing (dissoc extra k)))

      clojure.lang.IPersistentCollection
      (count [_] (count (snap)))
      (cons [_ o] (conj (snap) o))
      (empty [_] (array-map))
      (equiv [_ o] (= (snap) o))

      clojure.lang.Seqable
      (seq [_] (seq (snap)))

      Iterable
      (iterator [_] (.iterator ^Iterable (snap))))))

(defn- live-map? [val] (instance? sdk.test.omni.ILiveMap val))


;; --------------------------------------------------------------------------
;; Decision 1: the converters.
;; --------------------------------------------------------------------------

(defn ->struct
  "omni's immutable model -> the SDK's mutable one.

   ABSENT becomes struct's own NOARG (decision 2). A live view is unwrapped to
   the very map it reads, so a subject handed a contextified ctx mutates the
   node omni will match against (decision 3). Anything that is neither an omni
   map nor an omni list crosses as a leaf (decision 4)."
  [val]
  (cond
    (u/isabsent val) vs/NOARG
    (nil? val) nil
    (live-map? val) (live-backing val)
    (live? val) val
    (map? val) (let [out (LinkedHashMap.)]
                 (doseq [[k v] val] (.put out (str k) (->struct v)))
                 out)
    (vector? val) (let [out (ArrayList.)]
                    (doseq [v val] (.add out (->struct v)))
                    out)
    :else val))


(defn ->omni
  "The SDK's mutable model -> omni's immutable one.

   Clojure's own maps and vectors are already omni values, but a clojure map
   is ALSO a java.util.Map and a clojure vector a java.util.List, so the
   persistent branches must be tested first. `seen` guards a cyclic mutable
   node: struct's nodes are trees, but an SDK object reachable from one is not
   necessarily, and a converter that hangs is worse than one that truncates."
  ([val] (->omni val (IdentityHashMap.)))
  ([val ^IdentityHashMap seen]
   (cond
     (identical? vs/NOARG val) u/ABSENT
     (nil? val) nil
     (live-map? val) val
     (live? val) val
     (map? val) (reduce (fn [out k] (assoc out (str k) (->omni (get val k) seen)))
                        (array-map)
                        (keys val))
     (vector? val) (mapv #(->omni % seen) val)

     (instance? Map val)
     (if (.containsKey seen val)
       (array-map)
       (do (.put seen val true)
           (let [out (reduce (fn [out k] (assoc out (str k) (->omni (.get ^Map val k) seen)))
                             (array-map)
                             (vec (.keySet ^Map val)))]
             (.remove seen val)
             out)))

     (instance? List val)
     (if (.containsKey seen val)
       []
       (do (.put seen val true)
           (let [out (mapv #(->omni % seen) (vec val))]
             (.remove seen val)
             out)))

     :else val)))


;; --------------------------------------------------------------------------
;; Loading the corpus.
;; --------------------------------------------------------------------------

;; THE CORPUS IS PARSED BY THE SDK'S OWN JSON READER, not omni's.
;;
;; omni's readers model every JSON number as a double - the clojure port's
;; `json/parse` calls Double/parseDouble unconditionally, and its java peer
;; documents `number -> Double`. On most ports that is invisible, because
;; their struct has one numeric type. Clojure's does not: `typify` answers
;; T_integer for a Long and T_decimal for a Double, and `validate` prints
;; "integer" or "decimal" from it. Read through omni's parser, every `1` in
;; the corpus arrives as 1.0 and fourteen struct sections fail on the word
;; "decimal" - not because the SDK is wrong, but because the spec was
;; re-typed on the way in.
;;
;; omni's own `make-runner` takes an ALREADY-PARSED spec, so the fix needs no
;; change to the vendored files (which are read-only anyway): parse with the
;; reader the SDK itself uses, then convert into omni's model. Longs survive
;; the round trip, and omni compares numbers by value (`==`, and `numstr`
;; renders 5.0 as 5), so nothing downstream notices.
;;
;; The one distinction this cannot restore is an authored `1.0`: JSON does not
;; carry it and neither reader can. The corpus has no such literal today.
;; Upstream follow-up: omni's clojure json.clj could parse an integral literal
;; with no '.'/'e' as a Long, as the SDK's own reader does.
(defn load-spec [path]
  (when-not (.exists (java.io.File. ^String path))
    (throw (ex-info (str "omni: cannot read spec: " path) {})))
  (->omni (core/json-parse (slurp path))))


;; --------------------------------------------------------------------------
;; The subject boundary.
;; --------------------------------------------------------------------------

;; Decision 2a: the execution counter. omni calls the wrapped subject exactly
;; once per entry, and nothing else calls it at all, so the delta across a set
;; is the number of entries the ENGINE ran. Every corpus count in this port is
;; read from here.
(def ^:private subject-calls (atom 0))

(defn subject-call-count
  "Subject invocations since this process started."
  []
  @subject-calls)


;; Decision 2: does `f` have a genuine zero-argument arm?
;;
;; Read off the function object, not discovered by calling it. A
;; `(catch ArityException ...)` around the zero-arity call cannot tell a
;; subject that has NO zero-arity arm from one whose zero-arity arm raised an
;; ArityException from its own body - and for the second, catching it retries
;; the subject with nil (after any side effect of the first, partial call has
;; already landed) and reports the retry as a pass. A real crash becomes a
;; green entry.
;;
;; A fixed arity is a declared `invoke` of no parameters on the function's OWN
;; class - declared, never inherited, because `clojure.lang.AFn` declares an
;; `invoke()` that exists only to raise ArityException. A variadic `fn` is a
;; RestFn: its fixed arms are declared the same way, and its variadic arm
;; accepts no arguments exactly when it requires none.
(defn- accepts-no-args? [f]
  (boolean
   (or (some (fn [^java.lang.reflect.Method m]
               (and (= "invoke" (.getName m))
                    (zero? (alength (.getParameterTypes m)))))
             (.getDeclaredMethods ^Class (class f)))
       (and (instance? clojure.lang.RestFn f)
            (zero? (.getRequiredArity ^clojure.lang.RestFn f))))))


;; omni calls a subject with ONE argument: the resolved argument VECTOR. The
;; SDK's subjects are ordinary functions of their arguments, so the vector is
;; applied - which is also what keeps arity honest for decision 2.
;;
;; The return is `[args result]`, omni's `:runsetflags-args` shape: clojure's
;; data is immutable, so handing the arguments back is the only way a
;; `match: {args: ...}` can see what a subject did with them.
(defn- wrap-subject [subject]
  (when (ifn? subject)
    (fn [args]
      ;; Counted BEFORE the call, so an entry whose subject throws - which is
      ;; a pass when the entry expects it - still counts as executed.
      (swap! subject-calls inc)
      (let [converted (mapv ->struct args)
            ;; Decision 2: a lone ABSENT is "no argument at all", so the call
            ;; carries no argument. A subject without that arity gets nil.
            noarg? (and (= 1 (count args)) (u/isabsent (first args)))
            result (if noarg?
                     (if (accepts-no-args? subject) (subject) (subject nil))
                     (apply subject converted))]
        [(mapv ->omni converted) (->omni result)]))))


;; `makeContext` -> `:make-context`: corpus subject names are camelCase, the
;; SDK utility is a kebab-case keyword map.
(defn- utility-key [name]
  (-> (str name)
      (str/replace #"([a-z0-9])([A-Z])" "$1-$2")
      (str/lower-case)
      (keyword)))


;; The live SDK as an omni provider. Marked live so the converters never walk
;; it: omni attaches it to every contextified ctx under "client".
(defn- sdk-provider [sdk]
  (let [utility (when sdk (core/get-utility sdk))]
    (mark-live
     {:sdk sdk
      :utility utility

      ;; The corpus suites hand most subjects in per-set; this is the by-name
      ;; fallback omni uses when they do not.
      :subject (fn [name]
                 (when utility
                   (wrap-subject (get (deref utility) (utility-key name)))))

      ;; A DEF.client entry becomes another test SDK, rewrapped the same way.
      :client (fn [options]
                (sdk-provider (client/test-sdk nil (->struct options))))

      ;; Decision 3: omni keeps this return value as BOTH args[0] and
      ;; entry.ctx, so making it a live view over one mutable map is what
      ;; makes `match: {ctx: ...}` see a utility's writes.
      :contextify (fn [val]
                    (if (u/ismap val)
                      (live-map (->struct val) (array-map))
                      val))

      ;; Client options may reference the runner store.
      :inject (fn [options store]
                (let [opts (->struct options)]
                  (vs/inject opts (->struct store))
                  (->omni opts)))})))


;; --------------------------------------------------------------------------
;; The runner.
;; --------------------------------------------------------------------------

(defn make-runner
  "Vendored omni's `make-runner`, in the shape the corpus suites use.

   `specref` is a path to the shared corpus, or an already-parsed spec in
   omni's own value model (which keeps the smoke test free of fixture files)."
  ([specref] (make-runner specref nil))
  ([specref sdk]
   (let [provider (if sdk (sdk-provider sdk) (mark-live {}))
         spec (if (string? specref) (load-spec specref) specref)
         omnirunner (runner/make-runner spec provider)]
     (fn make-runpack
       ([name] (make-runpack name nil))
       ([name store]
        (let [runpack (omnirunner name store)
              ;; Always the `-args` entry point: the plain one drops the
              ;; argument writeback, and clojure has no other channel for it.
              ;;
              ;; RETURNS THE NUMBER OF ENTRIES THE ENGINE RAN (decision 2a),
              ;; measured at the subject wrapper. The suites compare it with
              ;; the set they handed in, so an engine that never called the
              ;; subject reports 0 rather than the set's own size.
              runsetflags (fn [testspec flags testsubject]
                            (let [before @subject-calls]
                              ((:runsetflags-args runpack)
                               testspec flags (wrap-subject testsubject))
                              (- @subject-calls before)))]
          {:spec (:spec runpack)
           :runset (fn [testspec testsubject] (runsetflags testspec {} testsubject))
           :runsetflags runsetflags
           :subject (:subject runpack)
           :client provider
           :provider provider}))))))


(defn null-modifier
  "Convert NULLMARK sentinels back into real nulls.

   NOT a delegation to omni's `nullmodifier`, deliberately: omni's RETURNS the
   replacement, while struct's `inject` passes this as its `modify` hook and
   expects it to MUTATE `parent[key]` in place."
  [val key parent & _]
  (cond
    (= NULLMARK val) (vs/setprop parent key nil)
    (string? val) (vs/setprop parent key (str/replace val NULLMARK "null"))))

File: ts/project/.sdk/tm/lua/test/omni.lua
-- ProjectName SDK corpus test runner: vendored @voxgig/omni driven through
-- its NATIVE API (`makeRunner(specref, provider)`), presented to the corpus
-- tests in the struct-runner shape they already use (`R.spec`, `R.runset`,
-- `R.runsetflags`, `R.client`). No compat shim is vendored: the adapter
-- below IS the whole bridge, per language, per the vendor-tag rollout
-- (docs/design/vendor-tag-rollout.md, Decision 4). It is the lua peer of
-- tm/ts/test/omni.ts / tm/py/test/omni.py, and a port of upstream omni's
-- own lua struct shim (voxgig/omni lua/compat/struct.lua) onto the REAL
-- generated SDK.
--
-- Four local decisions, all required:
--
-- 1. SPEC PATH. omni's own spec loading expects a usable path. A relative
--    path is absolutized against THIS module's directory (test/), so the
--    existing '../../.sdk/test/test.json' constant keeps working verbatim
--    wherever the suite is run from.
--
-- 2. PROVIDER DELEGATION. Corpus-driven contexts get `ctx.client` set to
--    the runner's provider (omni overwrites it on any ctx/args map entry).
--    A five-hook provider object HIDES the live SDK from the generated
--    utilities that reach through it - prepare_auth and prepare_headers via
--    `ctx.client:options_map()`, feature_add appending to client.features.
--    So the provider here is a READ-THROUGH view of the live SDK instance:
--    a table holding the omni hooks whose metatable delegates reads to the
--    SDK (`__index`) and routes writes onto it (`__newindex`) - the lua
--    spelling of ts's prototype delegation. (Upstream omni#56 tracks
--    giving the stock provider the same shape.)
--
-- 3. TWO VALUE MODELS, converted at the subject boundary. omni marks maps
--    and lists by METATABLE IDENTITY and models null/absent as sentinel
--    tables (a lua table cannot hold nil); the SDK and the vendored struct
--    speak dkjson's `__jsontype` convention with plain lua nil. Neither
--    recognises the other, so `wrapsubject` converts every argument on the
--    way in (tostruct) and the result on the way out (toomni), splicing
--    argument mutations back where omni's `match` can see them. This is
--    the port of upstream's compat shim, with ONE deliberate change: a
--    table that belongs to neither model (a live SDK object - the
--    provider, a Context, the utility) passes through BY REFERENCE as a
--    leaf, so subjects keep real object identity through `ctx.client`.
--
-- 4. NO-VALUE ARITY (the zero-argument correction six upstream compat
--    shims carry). An entry with no `in`/`args`/`ctx` must reach the
--    subject as NO value, not as one null. omni models absence as an
--    explicit ABSENT sentinel; the vendored struct's own NOVAL sentinel
--    (upstream struct/lua 0.1.1) is what an absent argument becomes, so
--    `typify()` answers T_noval where `typify(nil)` answers T_null. A
--    port without NOVAL gets trailing absent arguments TRIMMED instead,
--    which `select('#', ...)` reads as a zero-arity call.
--
-- THE VENDORED LUA PORT AND THE omni#54 RUNNER FIXES, checked per the
-- rollout: lua's `match` already reads its base directly (no clone), its
-- `errmessage` already keeps a message-bearing table's message (the SDK's
-- error type answers through __tostring), and its walkers recurse only
-- into omni-marked tables, so live cyclic objects are leaves to them. No
-- resolver workaround is needed at this tag; decision 3's leaf rule is
-- what keeps it that way.

local u = require("test.vendor.omni.util")
local Runner = require("test.vendor.omni.runner")
local vs = require("utility.struct.struct")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)") or "./"

local M = {}

M.NULLMARK = u.NULLMARK
M.UNDEFMARK = u.UNDEFMARK
M.EXISTSMARK = u.EXISTSMARK
M.JSON_NULL = u.NULL
M.ABSENT = u.ABSENT

-- omni's own model constructors and error type, re-exported for the smoke
-- test (an in-memory spec must be built in omni's value model).
M.map = u.map
M.list = u.list
M.OmniError = Runner.OmniError
M.isomnierror = Runner.isomnierror

--- The two value models, and the translation between them (decision 3).

local ARRAYMT = { __jsontype = "array" }
local OBJECTMT = { __jsontype = "object" }

--- omni's model -> struct's/SDK's.
---
--- Returns `(value, kind)` where kind is 'value', 'null' or 'absent'.
--- Three states, not two: omni's NULL and ABSENT both become a lua `nil`,
--- but they are NOT interchangeable at the call boundary - an absent
--- argument becomes the port's no-value, a null stays nil.
local function tostruct(val, seen)
  if u.NULL == val then
    return nil, "null"
  end
  if u.ABSENT == val then
    return nil, "absent"
  end
  if "table" ~= type(val) then
    return val, "value"
  end

  -- Decision 3's leaf rule: a table that is not omni-marked is a LIVE
  -- object (the provider, a Context, the utility). It crosses by
  -- reference, keeping identity and methods.
  local islist = u.islist(val)
  if not islist and not u.ismap(val) then
    return val, "value"
  end

  seen = seen or {}
  if nil ~= seen[val] then
    return seen[val], "value"
  end

  local out = setmetatable({}, islist and ARRAYMT or OBJECTMT)
  seen[val] = out

  if islist then
    for index = 1, #val do
      local entry, kind = tostruct(val[index], seen)
      -- A null INSIDE a list has to stay a slot, or the list shortens -
      -- and the slot is the STRING "null", which is what struct/lua's own
      -- runner put there and what the port is written against ("Preserve
      -- null in arrays as the string "null" to avoid nil holes").
      if "value" == kind then
        out[index] = entry
      else
        out[index] = "null"
      end
    end
  else
    for key, entry in pairs(val) do
      local converted, kind = tostruct(entry, seen)
      if "value" == kind then
        out[key] = converted
      end
    end
  end

  return out, "value"
end

--- Convert a value from omni's model into the SDK's (nulls become lua
--- nil), re-exported for test files that read spec sections directly -
--- DEF.setup options handed to `sdk.test`, corpus fixtures cloned before
--- driving a subject by hand.
function M.tostruct(val)
  local out = tostruct(val)
  return out
end

--- struct's/SDK's model -> omni's.
local function toomni(val, seen)
  if "table" ~= type(val) then
    return val
  end
  if u.NULL == val or u.ABSENT == val then
    return val
  end
  -- Already omni-marked: a subject handed its input straight back.
  if u.ismap(val) or u.islist(val) then
    return val
  end

  -- Struct marks with __jsontype; a live object carries some OTHER
  -- metatable and stays a leaf (decision 3). An unmarked plain table is
  -- classified the way struct/lua's own `islist` does it: at least one
  -- key, and the numeric keys are 1..n with nothing else (`count > 0`
  -- matters - an unmarked EMPTY table is a MAP to struct, and calling it
  -- a list turns every `{}` that comes back through merge or clone into
  -- `[]`).
  local mt = getmetatable(val)
  local jsontype = mt and mt.__jsontype
  if nil ~= mt and nil == jsontype then
    return val
  end

  local islist
  if "array" == jsontype then
    islist = true
  elseif "object" == jsontype then
    islist = false
  else
    local count, max = 0, 0
    islist = true
    for key in pairs(val) do
      if "number" ~= type(key) then
        islist = false
        break
      end
      count = count + 1
      if key > max then
        max = key
      end
    end
    islist = islist and 0 < count and max == count
  end

  seen = seen or {}
  if nil ~= seen[val] then
    return seen[val]
  end

  local out = islist and u.list({}) or u.map({})
  seen[val] = out

  if islist then
    for index = 1, #val do
      out[index] = toomni(val[index], seen)
    end
  else
    for key, entry in pairs(val) do
      out[key] = toomni(entry, seen)
    end
  end

  return out
end

--- Replace the CONTENTS of an omni table with another's, keeping the
--- original table's identity and its map/list metatable. Used to write an
--- argument mutation back where omni's `match` can see it.
local function splice(into, from)
  local keys = {}
  for key in pairs(into) do
    keys[#keys + 1] = key
  end
  for _, key in ipairs(keys) do
    into[key] = nil
  end
  for key, value in pairs(from) do
    into[key] = value
  end
end

--- A subject that speaks the SDK's value model on the way in and omni's on
--- the way out. Arity is preserved exactly: `select('#', ...)` is what
--- tells an entry with no `in` (zero arguments) from one with `in: null`
--- (one) - decision 4.
local function wrapsubject(subject, noval)
  if "function" ~= type(subject) then
    return nil
  end
  return function(...)
    local count = select("#", ...)
    local args = table.pack(...)
    local converted, presence = {}, {}
    for index = 1, count do
      local value, kind = tostruct(args[index])
      -- NOT `kind and value or nil`: when `value` is `false` that whole
      -- expression is `nil`, so every `in: false` entry arrived as absent.
      presence[index] = ("absent" ~= kind)
      if "value" == kind then
        converted[index] = value
      end
    end

    -- An absent argument becomes the port's own no-value, so `typify()`
    -- can answer T_noval where `typify(nil)` answers T_null. Without a
    -- sentinel there is nothing to send, so trailing absents shorten the
    -- call instead and the port sees a zero-arity call.
    if nil ~= noval then
      for index = 1, count do
        if not presence[index] then
          converted[index] = noval
        end
      end
    else
      while 0 < count and not presence[count] do
        count = count - 1
      end
    end

    local result = subject(table.unpack(converted, 1, count))

    -- SDK utilities MUTATE the value they are given (result_body rewrites
    -- ctx.result in place), and the corpus checks it via `match`. But
    -- `tostruct` built a copy, so the mutation landed on the copy - splice
    -- the result back into omni's own table, keeping its identity so its
    -- map/list marking survives. Only a converted COPY is written back: a
    -- leaf that crossed by reference (decision 3) mutated in place
    -- already, and self-splicing it would wipe it.
    for index = 1, count do
      local original, mutated = args[index], converted[index]
      if
        "table" == type(original)
        and "table" == type(mutated)
        and original ~= mutated
        and (u.ismap(original) or u.islist(original))
      then
        splice(original, toomni(mutated))
      end
    end

    -- The port's no-value is omni's absent, explicitly.
    if nil ~= noval and noval == result then
      return u.ABSENT
    end
    -- A plain `nil` back from the SDK is ABSENT, not NULL. The port has
    -- one `nil` for both, and the corpus decides which way: measured
    -- upstream over the struct corpus, reading it as null costs 43
    -- entries, reading it as absent costs 6 (named in the struct suite's
    -- drop guards).
    if nil == result then
      return u.ABSENT
    end
    return toomni(result)
  end
end

--- `makeContext` -> `make_context`: corpus subject names are camelCase,
--- the SDK utility is snake_case.
local function snake(name)
  return (name:gsub("%u", function(c)
    return "_" .. c:lower()
  end))
end

--- Read `name` off the SDK's utility the way struct's runner does: the
--- utility itself (either spelling), then the struct utilities.
local function lookup(utility, name)
  if nil == utility then
    return nil
  end
  local found = utility[name]
  if nil == found then
    found = utility[snake(name)]
  end
  if nil == found and nil ~= utility.struct then
    found = utility.struct[name]
  end
  return found
end

--- Wrap the live SDK as an omni provider (decision 2): the omni hooks sit
--- in the table, everything else delegates to the SDK - reads through
--- `__index`, writes through `__newindex`.
local function sdkprovider(client)
  -- The LIVE utility, not the copy `get_utility()` hands out - the corpus
  -- suites reach the struct utilities through `client.utility().struct`
  -- (mirrors ts), so they are attached here once.
  local utility = client._utility
  if nil ~= utility and nil == utility.struct then
    utility.struct = vs
  end

  local noval = nil ~= utility and nil ~= utility.struct
    and utility.struct.NOVAL or nil

  local provider = {}

  provider.sdk = client

  provider.subject = function(name)
    return wrapsubject(lookup(utility, name), noval)
  end

  -- A DEF.client entry becomes another SDK instance - rewrapped with the
  -- same delegating shape, not a plain hook object.
  provider.client = function(options)
    local opts = tostruct(options)
    return sdkprovider(client.test(nil, opts))
  end

  -- omni's resolveargs installs `client` on a contextified map argument;
  -- the utility is added here so corpus contexts carry both (struct's
  -- runner handed its contexts the same pair). The ctx STAYS an omni map -
  -- the subject adapters materialise a real Context from it, and
  -- wrapsubject splices their writebacks into it for `match: {ctx: ...}`.
  provider.contextify = function(val)
    if nil ~= utility and nil ~= utility.contextify then
      val = utility.contextify(val)
    end
    if u.ismap(val) then
      rawset(val, "utility", utility)
    end
    return val
  end

  -- Client options may reference the runner store.
  provider.inject = function(options, store)
    if nil ~= utility and nil ~= utility.struct
      and nil ~= utility.struct.inject then
      utility.struct.inject(options, store)
    end
    return options
  end

  provider.utility = function()
    return utility
  end

  return setmetatable(provider, {
    __index = client,
    __newindex = function(_, key, value)
      client[key] = value
    end,
  })
end


--- struct's makeRunner(testfile, client) signature, backed by vendored
--- omni. Also accepts an already-parsed spec (omni's own capability, in
--- omni's value model), which keeps smoke tests free of fixture files.
function M.makeRunner(testfile, client)
  local specref = testfile
  if "string" == type(testfile) and "/" ~= testfile:sub(1, 1) then
    specref = _test_dir .. testfile
  end

  local provider = sdkprovider(client)
  local runner = Runner.makeRunner(specref, provider)

  local utility = client._utility
  local noval = nil ~= utility and nil ~= utility.struct
    and utility.struct.NOVAL or nil

  return function(name, store)
    local runpack = runner(name, store or u.map({}))

    -- Explicitly passed subjects speak the SDK's model too - the corpus
    -- tests hand most subjects in per-set rather than by name.
    local runsetflags = function(testspec, flags, testsubject)
      return runpack.runsetflags(testspec, flags, wrapsubject(testsubject, noval))
    end

    return {
      spec = runpack.spec,
      runset = function(testspec, testsubject)
        return runsetflags(testspec, {}, testsubject)
      end,
      runsetflags = runsetflags,
      subject = runpack.subject,
      -- The read-through provider: tests treat it as the SDK (its members
      -- all resolve), and omni hands it to corpus contexts as ctx.client.
      client = provider,
    }
  end
end

--- Convert NULLMARK sentinels back into real nulls.
---
--- NOT a delegation to omni's `nullmodifier`, deliberately: omni's
--- RETURNS the replacement value, while struct/lua's `inject` passes this
--- as its `modify` hook and expects it to MUTATE `parent[key]` in place.
--- It writes a real lua `nil` rather than omni's NULL sentinel, because
--- that is what struct's own modifier wrote and what the corpus expects
--- to come back out.
function M.nullModifier(val, key, parent)
  if M.NULLMARK == val then
    parent[key] = nil
  elseif "string" == type(val) then
    parent[key] = (val:gsub(M.NULLMARK, "null"))
  end
end

return M
